Metalama.Framework.Engine/Diagnostics/UserDiagnosticSink.cs

```diff
@@ -168,8 +169,44 @@ internal void Suppress( IEnumerable<ScopedSuppression> suppressions )
             }
         }
 
+        internal const string DeduplicationPropertyKey = "Metalama.Deduplication";
+
         public ImmutableUserDiagnosticList ToImmutable()
-            => new( this._diagnostics?.ToImmutableArray(), this._suppressions?.ToImmutableArray(), this._codeFixes?.ToImmutableArray() );
+        {
+            ImmutableArray<Diagnostic>? immutableDiagnostics = null;
+
+            var diagnostics = this._diagnostics;
+
+            if ( diagnostics != null )
+            {
+                HashSet<string>? deduplicationKeys = null;
```

Comment:
I think this should be deduplicated based on the diagnostic code AND the deduplication key.